craft_parts/overlays/chroot.py

```diff
@@ -0,0 +1,147 @@
+    """Execute a callable in a chroot environment.
+
+    :param path: The new filesystem root.
+    :param target: The callable to run in the chroot environment.
+    :param args: Arguments for target.
+    :param kwargs: Keyword arguments for target.
+
+    :returns: The target function return value.
+    """
+    logger.debug("[pid=%d] parent process", os.getpid())
+    parent_conn, child_conn = multiprocessing.Pipe()
+    child = multiprocessing.Process(
+        target=_runner, args=(Path(path), child_conn, target, args, kwargs)
+    )
+    _setup_chroot(path)
+    try:
+        child.start()
+        res, exc = parent_conn.recv()
+        child.join()
+    finally:
+        _cleanup_chroot(path)
+
+    if isinstance(exc, Exception):
+        raise exc
+
+    return res
+
+
+def _runner(
+    path: Path,
+    conn: Connection,
+    target: Callable,
+    args: Tuple,
+    kwargs: Dict,
+) -> None:
+    """Chroot to the execution directory and call the target function."""
+    logger.debug("[pid=%d] child process: target=%r", os.getpid(), target)
+    try:
+        logger.debug("[pid=%d] chroot to %r", os.getpid(), path)
+        os.chdir(path)
+        os.chroot(path)
+        res = target(*args, **kwargs)
+    except Exception as exc:  # pylint: disable=broad-except
+        conn.send((None, exc))
+        return
+
+    conn.send((res, None))
+
+
+def _setup_chroot(path: Path) -> None:
+    """Prepare the chroot environment before executing the target function."""
+    logger.debug("setup chroot: %r", path)
+    if sys.platform == "linux":
+        _setup_chroot_linux(path)
+
+
+def _cleanup_chroot(path: Path) -> None:
+    """Clean the chroot environment after executing the target function."""
+    logger.debug("cleanup chroot: %r", path)
+    if sys.platform == "linux":
+        _cleanup_chroot_linux(path)
+
+
+_Mount = namedtuple("_Mount", ["fstype", "src", "mountpoint", "option"])
+
+_linux_mounts: List[_Mount] = [
+    _Mount(None, "/etc/resolv.conf", "/etc/resolv.conf", "--bind"),
+    _Mount("proc", "proc", "/proc", None),
+    _Mount("sysfs", "sysfs", "/sys", None),
+    _Mount(None, "/dev", "/dev", "--bind"),
+    _Mount("tmpfs", "tmpfs", "/dev/shm", None),
+]
+
+
+def _setup_chroot_linux(path: Path) -> None:
+    for entry in _linux_mounts:
+        args = []
+        if entry.option:
+            args.append(entry.option)
+        if entry.fstype:
+            args.append(f"-t{entry.fstype}")
+
+        mountpoint = path / entry.mountpoint.lstrip("/")
+
+        # Some images (such as cloudimgs) symlink ``/etc/resolv.conf`` to
+        # ``/run/systemd/resolve/stub-resolv.conf``. We want resolv.conf to be
+        # a regular file to bind-mount the host resolver configuration on.
+        #
+        # There's no need to restore the file to its original conditions because
+        # this operation happens on a temporary filesystem layer.
+        if entry.src == "/etc/resolv.conf":
```

Comment:
I don't like this special case here. I understand that resolv.conf may need extra work, but why not make it generic? Like adding a parameter in `_Mount` above?

IOW, this looks like instead of current lines 111 and 112 (which I like), you'd have there:

```
if entry.src == "/etc/resolv.conf" or entry.src == "/dev":
    args.append("--bind")
```

Reply:
This is a special case handling that can be done separately, keeping the mount loop uniform. Changed the code to perform special case verification before entering the mount list processing.